tests/rptest/services/redpanda_installer.py

```diff
@@ -0,0 +1,231 @@
+        return f"{RedpandaInstaller.INSTALLER_ROOT}/{version}"
+
+    @staticmethod
+    def wait_for_async_ssh(logger, ssh_out_per_node, log_msg):
+        """
+        Waits for each SSHOutputIter to complete.
+        """
+        for node in ssh_out_per_node:
+            logger.debug(f"{log_msg} for {node.account.hostname}")
+            for l in ssh_out_per_node[node]:
+                logger.debug(l)
+
+    def __init__(self, redpanda):
+        """
+        Constructs an installer for the given RedpandaService.
+        """
+        self._started = False
+        self._redpanda = redpanda
+        self._installed_per_node = dict()
+
+        # Keep track if the original install path is /opt/redpanda is used, as
+        # is the case for package-deployed clusters. Since the installer uses
+        # this directory, we'll need to be mindful not to mess with the
+        # original binaries.
+        rp_install_path_root = self._redpanda._context.globals.get(
+            "rp_install_path_root", None)
+        self._head_backed_up = rp_install_path_root == "/opt/redpanda"
+
+    def start(self):
+        """
+        Validates that all nodes in the service have installed the same
+        version, and initializes test-wide state, like the list of released
+        versions.
+        """
+        if self._started:
+            return
+
+        initial_version = None
+        nodes = self._redpanda.nodes
+
+        # Verify that the installations on each node match.
+        for node in nodes:
+            vers = self._redpanda.get_version(node)
+            if initial_version == None:
+                initial_version = vers
+            assert initial_version == vers, \
+                f"Mismatch version {node.account.hostname} has {vers}, {nodes[0].account.hostname} has {initial_version}"
+
+        # Clean up the installer root directory so we start out clean.
+        for node in nodes:
+            if node.account.exists(RedpandaInstaller.INSTALLER_ROOT):
+                node.account.remove(f"{RedpandaInstaller.INSTALLER_ROOT}/*",
+                                    allow_fail=True)
+            else:
+                node.account.mkdir(RedpandaInstaller.INSTALLER_ROOT)
+
+        # Now that we're at a sane starting point, set up our install path for
+        # ease of jumping between versions.
+        ssh_setup_head_per_node = dict()
+        head_root_path = RedpandaInstaller.root_for_version(
+            RedpandaInstaller.HEAD)
+        rp_install_path_root = self._redpanda._context.globals.get(
+            "rp_install_path_root", None)
+        for node in nodes:
+            # For simplicity's sake, always end up with binaries at
+            # 'head_root_path', so we can continue to use root_for_version() to
+            # reference the head root.
+            head_cmd = ""
+            if self._head_backed_up:
+                head_cmd = f"mv /opt/redpanda {head_root_path}"
+            else:
```

Comment:
Until we have package management, let's not mess with the underlying files. There's something to be said about the reliability of tests, e.g. if a test crashes, a session ends, etc., we end up in a botched state where the state doesn't match what the package manager expects. Consider not running these tests if running with dedicated nodes.

Reply:
I agree in general there's room to improve this soon with the addition of package management, but I've been going back and forth about support here. It is important to have upgrade capabilities in CDT, even if it's stitched together as it currently is, e.g. for the sake of reproducing bugs that only happen during upgrades at scale. Yes, package management would add significantly more legitimacy to such tests, but I don't see package installation taking all that much test coverage away from the Redpanda binary.

Offline you mentioned some concerns around when you're running CDT and the connection is disrupted, and you're left with botched state that might conflict with the expectations of the package manager, e.g. because the teardown hook failed to run. I'm thinking a compromise here would be to proceed with this linked, but:
1. always use an installer when using `RedpandaService`, and
2. when initializing the `RedpandaService`, check if `/opt/redpanda` is a symlink and if there exist any `redpanda_install/head` binaries. If so, we swap them into `/opt/redpanda`

Does that quell at least your operational concerns of using symlinks?